tests/defences/detector/poison/test_activation_defence.py

```diff
@@ -28,285 +28,290 @@
-        self.assertNotEqual(sum_dis_gen, sum_sz_gen, msg="This is very unlikely to happen... there may be an error")
-
-    def test_plot_clusters(self):
-        self.defence.detect_poison(nb_clusters=2, nb_dims=10, reduce="PCA")
-        self.defence_gen.detect_poison(nb_clusters=2, nb_dims=10, reduce="PCA")
-        self.defence.plot_clusters(save=False)
-        self.defence_gen.plot_clusters(save=False)
-
-    def test_pickle(self):
-
-        # Test pickle and unpickle:
-        filename = "test_pickle.h5"
-        ActivationDefence._pickle_classifier(self.classifier, filename)
-        loaded = ActivationDefence._unpickle_classifier(filename)
-
-        np.testing.assert_equal(self.classifier._clip_values, loaded._clip_values)
-        self.assertEqual(self.classifier._channels_first, loaded._channels_first)
-        self.assertEqual(self.classifier._use_logits, loaded._use_logits)
-        self.assertEqual(self.classifier._input_layer, loaded._input_layer)
-
-        ActivationDefence._remove_pickle(filename)
-
-    def test_fix_relabel_poison(self):
-        (x_train, y_train), (_, _), (_, _) = self.mnist
-        x_poison = x_train[:100]
-        y_fix = y_train[:100]
-
-        test_set_split = 0.7
-        n_train = int(len(x_poison) * test_set_split)
-        x_test = x_poison[n_train:]
-        y_test = y_fix[n_train:]
-
-        predictions = np.argmax(self.classifier.predict(x_test), axis=1)
-        ini_miss = 1 - np.sum(predictions == np.argmax(y_test, axis=1)) / y_test.shape[0]
-
-        improvement, new_classifier = ActivationDefence.relabel_poison_ground_truth(
-            self.classifier,
-            x_poison,
-            y_fix,
-            test_set_split=test_set_split,
-            tolerable_backdoor=0.01,
-            max_epochs=5,
-            batch_epochs=10,
-        )
-
-        predictions = np.argmax(new_classifier.predict(x_test), axis=1)
-        final_miss = 1 - np.sum(predictions == np.argmax(y_test, axis=1)) / y_test.shape[0]
-
-        self.assertEqual(improvement, ini_miss - final_miss)
-
-        # Other method (since it's cross validation we can't assert to a concrete number).
-        improvement, _ = ActivationDefence.relabel_poison_cross_validation(
-            self.classifier, x_poison, y_fix, n_splits=2, tolerable_backdoor=0.01, max_epochs=5, batch_epochs=10
-        )
-        self.assertGreaterEqual(improvement, 0)
-
-    def test_visualizations(self):
-        # test that visualization doesn't error in grayscale and RGB settings
-        (x_train, _), (_, _), (_, _) = self.mnist
-        self.defence.visualize_clusters(x_train)
-
-        x_train_rgb = convert_to_rgb(x_train)
-        self.defence.visualize_clusters(x_train_rgb)
-
-
-if __name__ == "__main__":
-    unittest.main()
+            n_dp += len(clusters_by_class[i])
+        assert len(x_train) == n_dp
+
+
+@pytest.mark.skipMlFramework("non_dl_frameworks", "pytorch", "mxnet", "tensorflow")
```

Comment:
Does Activation Defense also support TensorFlow and PyTorch? If we change the test classifier to `image_dl_estimator` we should be able to test other frameworks.

Reply:
@ebubae Did you notice this comment?